packages/core/src/components/autosuggest/autosuggest.tsx

```diff
@@ -0,0 +1,302 @@
+})
+export class Autosuggest {
+  @Prop()
+  fetchSuggestions: (value: string) => Promise<Array<Suggestion>> = () =>
+    Promise.resolve([]);
+
+  @Prop()
+  suggestOnFocus: boolean = false;
+
+  @Element()
+  host!: HTMLElement;
+
+  @State()
+  suggestions: Array<SuggestionState> = [];
+
+  selectedIndex: number = -1;
+
+  terms: string[] = [];
+
+  @State()
+  showSuggestions: boolean = false;
+
+  input!: HTMLInputElement;
+
+  listboxId: string = v4();
+
+  inputId: string = v4();
+
+  labelId: string = v4();
+
+  listbox: HTMLUListElement | undefined;
+
+  debouncedFetchSuggestions = debounce(
+    (terms: string[]) =>
+      this.fetchSuggestions(terms.join(" "))
+        .then((result) => {
+          this.suggestions = result.map((suggestion, index) => ({
+            value: suggestion.value,
+            type: suggestion.type,
+            selected: false,
+            id: `${index}-${this.inputId}`,
+          }));
+          this.terms = terms;
+          this.openSuggestions();
+        })
+        .catch(() => {
+          this.closeSuggestions();
+          this.suggestions = [];
+        }),
+    200
+  );
+
+  onInput = (event: Event) => {
+    if (!(event.target instanceof HTMLInputElement)) {
+      throw new Error("event.target not a instance of HTMLInputElement");
+    }
+
+    this.debouncedFetchSuggestions(
+      `${event.target.value}`.match(/(\S+)/g) ?? []
+    );
+
+    this.input.value = event.target.value;
+  };
+
+  onFocusIn = () => {
+    if (this.suggestOnFocus) {
+      this.openSuggestions();
+    }
+  };
+
+  onDocumentClick = (event: MouseEvent) => {
+    if (
+      this.showSuggestions &&
+      this.listbox &&
+      event.target instanceof Node &&
+      !this.listbox.contains(event.target) &&
+      this.input != event.target
+    ) {
+      this.closeSuggestions();
+    }
+  };
+
+  connectedCallback() {
+    const input = this.host.querySelectorAll('input[type="text"]')[0];
+    if (!(input instanceof HTMLInputElement)) {
+      throw new ReferenceError("Mandatory text input not found");
+    }
+
+    this.input = input;
+    if (input.id) {
+      this.inputId = input.id;
+    } else {
+      input.id = this.inputId;
+    }
+
+    if (!this.input.labels || this.input.labels.length < 1) {
+      throw new ReferenceError("Mandatory label for text input not found");
+    }
+
+    const label = this.input.labels[0];
+    if (label.id) {
+      this.labelId = label.id;
+    } else {
+      label.id = this.labelId;
+    }
+
+    this.input.setAttribute("autocomplete", "off");
+    this.input.setAttribute("aria-autocomplete", "list");
+    this.input.setAttribute("aria-controls", this.listboxId);
+    this.input.setAttribute("aria-activedescendant", "");
+    this.input.addEventListener("input", this.onInput);
+    this.input.addEventListener("keydown", this.onKeyDown);
+    this.input.addEventListener("focusin", this.onFocusIn);
+    window.document.addEventListener("click", this.onDocumentClick);
```

Comment:
Volgens mij kan je dit met de `@Listen()` decorator fraaier oplossen:

```
@Listen('click', { target: 'document' })
onDocumentClick(event: MouseEvent) {
  // [..]
}
```

Maar ik rammel dit uit de losse pols, geen garanties ;-) Stencil heeft er in ieder geval iets op bedacht: https://stenciljs.com/docs/events#listen-decorator

Reply:
aangepast